### PR TITLE
feat: provide gateway-rest in all components

### PR DESCRIPTION
**Pull request by romansmirnov**
## Description

Exposes the Camunda REST API on all nodes. Meaning, in addition to the broker and gateway, Operate, and Tasklist also bootstrap the Camunda REST API.

* The Camunda REST API can be disabled with `camunda.rest.enabled=false` (by default, the Camunda REST API is enabled).
* Any node that bootstraps the Camunda REST API connects to the Zeebe Cluster by configuring the `zeebe.gateway.cluster.*` properties (like initial contact points, host, and port). That way, when the node starts, it will join the Zeebe Cluster and retrieve the topology.

## Related issues

related to #19076

### CODE REVIEW

File: dist/src/main/config/broker.yaml.template [hunk 1]
```diff
@@ -30,6 +30,10 @@
 # Relative paths are resolved relative to the installation directory of the broker.
 # ----------------------------------------------------
 
+# Enable the Camunda REST API to start on broker startup.
+# This setting can also be overridden using the environment variable CAMUNDA_REST_ENABLED.
+# camunda.rest.enabled: false
```
Comment (Zelldon) [inline dist/src/main/config/broker.yaml.template:35]:
```suggestion
# camunda.rest.enabled: true 
```
We normally show the default

File: dist/src/main/java/io/camunda/application/StandaloneOperate.java [hunk 2]
```diff
@@ -37,7 +39,11 @@ public static void main(final String[] args) {
 
     final var standaloneOperateApplication =
         MainSupport.createDefaultApplicationBuilder()
-            .sources(OperateModuleConfiguration.class, WebappsModuleConfiguration.class)
+            .sources(
+                ApplicationModuleConfiguration.class,
+                CommonsModuleConfiguration.class,
```
Comment (Zelldon) [inline dist/src/main/java/io/camunda/application/StandaloneOperate.java:44]: 💭 Can these both consolidated as well? 🤔 As we reference them all the time?

File: dist/src/main/java/io/camunda/application/configuration/BrokerBasedConfiguration.java [hunk 3]
```diff
@@ -68,7 +72,7 @@ public SchedulerConfiguration schedulerConfiguration() {
     return new SchedulerConfiguration(cpuThreads, ioThreads, metricsEnabled, "Broker", nodeId);
   }
 
-  @ConditionalOnProperty(prefix = "zeebe.broker.gateway", name = "enable", havingValue = "false")
+  @ConditionalOnProperty(prefix = "camunda.rest", name = "enabled", havingValue = "false")
```
Comment (Zelldon) [inline dist/src/main/java/io/camunda/application/configuration/BrokerBasedConfiguration.java:75]: ❓ Does it mean we are not disabling it when we disable the embedded gateway? 🤔

File: dist/src/main/java/io/camunda/application/configuration/GatewayBasedConfiguration.java [hunk 4]
```diff
@@ -14,23 +14,90 @@
...
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Profile;
 
 @Configuration(proxyBeanMethods = false)
-public final class GatewayClusterConfiguration {
+@EnableConfigurationProperties(GatewayBasedProperties.class)
+@Profile("!broker")
+public final class GatewayBasedConfiguration {
```
Comment (Zelldon) [inline dist/src/main/java/io/camunda/application/configuration/GatewayBasedConfiguration.java:42]: ❓ How does this work with the embedded Gateway? I guess this is only for standalone gateway necessary?

File: dist/src/main/java/io/camunda/commons/CommonsModuleConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
+import org.springframework.boot.context.properties.ConfigurationPropertiesScan;
+import org.springframework.context.annotation.ComponentScan;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration(proxyBeanMethods = false)
+@ComponentScan("io.camunda.commons")
+@ConfigurationPropertiesScan("io.camunda.commons")
+@EnableAutoConfiguration
+public class CommonsModuleConfiguration {}
```
Comment (Zelldon) [inline dist/src/main/java/io/camunda/commons/CommonsModuleConfiguration.java:19]: ❓ 💭 Could this be part of the ApplicationModule class?

File: dist/src/main/java/io/camunda/commons/clustering/AtomixClusterConfiguration.java [hunk 6]
```diff
@@ -27,6 +27,8 @@ public AtomixClusterConfiguration(final ClusterConfig config) {
 
   @Bean(destroyMethod = "stop")
   public AtomixCluster atomixCluster() {
-    return new AtomixCluster(config, Version.from(VersionUtil.getVersion()));
+    final var atomixCluster = new AtomixCluster(config, Version.from(VersionUtil.getVersion()));
+    atomixCluster.start();
```
Comment (Zelldon) [inline dist/src/main/java/io/camunda/commons/clustering/AtomixClusterConfiguration.java:31]: ❓ Where was this started before? For example in the broker bootstrap right? 🤔 How does this affect now the bootstrapping, is this an issue?

File: dist/src/main/java/io/camunda/commons/job/JobHandlerConfiguration.java [hunk 7]
```diff
@@ -27,8 +27,8 @@
 
 @Configuration(proxyBeanMethods = false)
 @ConditionalOnProperty(
-    prefix = "zeebe.broker.gateway",
-    name = "enable",
+    prefix = "camunda.rest",
+    name = "enabled",
     havingValue = "true",
```
Comment (Zelldon) [inline dist/src/main/java/io/camunda/commons/job/JobHandlerConfiguration.java:32]: ❌ I feel like we are breaking here something. Can't we support the old configs as well?

File: dist/src/main/java/io/camunda/zeebe/gateway/GatewayModuleConfiguration.java [hunk 8]
```diff
@@ -35,20 +37,23 @@
 @Configuration(proxyBeanMethods = false)
 @ComponentScan(
     basePackages = {
-      "io.camunda.commons",
```
Comment (Zelldon) [inline dist/src/main/java/io/camunda/zeebe/gateway/GatewayModuleConfiguration.java:38]: ❓ this was removed as we have now the CommonsConfiguration component right?

File: dist/src/main/java/io/camunda/zeebe/restore/RestoreApp.java [hunk 9]
```diff
@@ -21,11 +22,11 @@
...
 import org.springframework.boot.context.properties.ConfigurationPropertiesScan;
+import org.springframework.context.annotation.Import;
 
-@SpringBootApplication(
-    scanBasePackages = {"io.camunda.zeebe.restore", "io.camunda.zeebe.broker.shared"})
-@ConfigurationPropertiesScan(
-    basePackages = {"io.camunda.zeebe.broker.shared", "io.camunda.zeebe.restore"})
+@SpringBootApplication(scanBasePackages = {"io.camunda.zeebe.restore"})
+@ConfigurationPropertiesScan(basePackages = {"io.camunda.zeebe.restore"})
+@Import(value = {BrokerBasedConfiguration.class, WorkingDirectoryConfiguration.class})
```
Comment (Zelldon) [inline dist/src/main/java/io/camunda/zeebe/restore/RestoreApp.java:29]: ❓ The import allows us to remove the scan config?

File: zeebe/qa/util/src/main/java/io/camunda/zeebe/qa/util/cluster/TestSpringApplication.java [hunk 10]
```diff
@@ -27,7 +28,7 @@
 
 abstract class TestSpringApplication<T extends TestSpringApplication<T>>
     implements TestApplication<T> {
-  private final Class<?> springApplication;
+  private final List<Class<?>> springApplications;
```
Comment (Zelldon) [inline zeebe/qa/util/src/main/java/io/camunda/zeebe/qa/util/cluster/TestSpringApplication.java:31]: 😄 This is also part of https://github.com/camunda/camunda/pull/18964

File: zeebe/qa/util/src/main/java/io/camunda/zeebe/qa/util/cluster/TestCamundaApplication.java [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.zeebe.qa.util.cluster;
+
+import io.camunda.application.configuration.ApplicationModuleConfiguration;
+import io.camunda.commons.CommonsModuleConfiguration;
+import java.util.List;
+
+public abstract class TestCamundaApplication<T extends TestCamundaApplication<T>>
```
Comment (Zelldon) [inline zeebe/qa/util/src/main/java/io/camunda/zeebe/qa/util/cluster/TestCamundaApplication.java:14]: 💭 Maybe you can review https://github.com/camunda/camunda/pull/18964 then we can change it on the new camunda app 🤔

**Review by ChrisKujawa — COMMENTED** [review]
Thanks @romansmirnov as you know Spring is a lot of magic, which raises to me often lot of question so sorry for that :D But would be great if we can clarify them :)

File: dist/src/main/config/broker.yaml.template (hunk 1 above)
Comment (romansmirnov) [reply dist/src/main/config/broker.yaml.template:35]: Ah, I assumed that that specific configuration template shows how to configure the Broker without an embedded gateway, that's what it also saying at the top of the file:
```
# Zeebe Broker configuration file (without embedded gateway)
```

That's why I added it here and set it to `false`. Do you want me to change that?

File: dist/src/main/java/io/camunda/application/StandaloneOperate.java (hunk 2 above)
Comment (romansmirnov) [reply dist/src/main/java/io/camunda/application/StandaloneOperate.java:44]: Yes, I plan to do it separately also when having Tasklist and Identity included. Right now, we have too many fine-grained configuration classes, which can cause issues, because something might have been forgotten, etc. So my plan is once we have the Single Application with Tasklist and Identity, then I want to re-iterate this part and consolidate that.

File: dist/src/main/java/io/camunda/application/configuration/GatewayBasedConfiguration.java (hunk 4 above)
Comment (romansmirnov) [reply dist/src/main/java/io/camunda/application/configuration/GatewayBasedConfiguration.java:42]: This does not affect the embedded gateway. The embedded gateway is started only with profile `broker` (and when `zeebe.broker.gateway.enable` is set to `true`). Since the `broker` profile is used for the embedded gateway, the application will use the `BrokerBasedConfiguration` and configure the embedded gateway accordingly (as done previously).

The condition `!broker` assumes that if the broker is not started with the application, then the node acts "only" as a "gateway," exposing the Camunda API, aka v2 endpoints (in addition to potentially serving a respective UI).

To sum up, the application configuration happens either by
1. `BrokerBasedConfiguration`: using the `zeebe.broker.*` configurations, or
2. `GatweayBasedConfiguration`: using the `zeebe.gateway.*` configurations.

Based on that, commons modules (i.e., the actor scheduler, Atomix cluster, etc.) are started so that the Camunda API can be exposed on any node.

File: dist/src/main/java/io/camunda/commons/CommonsModuleConfiguration.java (hunk 5 above)
Comment (romansmirnov) [reply dist/src/main/java/io/camunda/commons/CommonsModuleConfiguration.java:19]: I plan to consolidate these configuration classes to avoid having too many fine-grained, as mentioned above. However, I would tackle it in a separate pull request as mentioned above.

File: dist/src/main/java/io/camunda/zeebe/gateway/GatewayModuleConfiguration.java (hunk 8 above)
Comment (romansmirnov) [reply dist/src/main/java/io/camunda/zeebe/gateway/GatewayModuleConfiguration.java:38]: Yep, I wanted to avoid adding that base package everywhere. That's why I moved to the `CommonsModuleConfiguration`, which will do the scanning there.

File: dist/src/main/java/io/camunda/zeebe/restore/RestoreApp.java (hunk 9 above)
Comment (romansmirnov) [reply dist/src/main/java/io/camunda/zeebe/restore/RestoreApp.java:29]: Yes, exactly. Previously, the `io.camunda.zeebe.broker` package had a `shared` package inside that was included as a base package in the `RestoreApp` so that Spring also includes these two configuration classes. I was messing around wit the `shared` package just to make the `RestoreApp` working, but eventually went this approach to keep it easy.

Coming back to your question, with the `@Import` annotation additional configuration classes can be included for the scanning.

File: zeebe/qa/util/src/main/java/io/camunda/zeebe/qa/util/cluster/TestSpringApplication.java (hunk 10 above)
Comment (romansmirnov) [reply zeebe/qa/util/src/main/java/io/camunda/zeebe/qa/util/cluster/TestSpringApplication.java:31]: I knew that, but I wanted to make the PR green 😅 Sure, I can review the other PR, then this part should become obsolete.

File: zeebe/qa/util/src/main/java/io/camunda/zeebe/qa/util/cluster/TestCamundaApplication.java (hunk 11 above)
Comment (romansmirnov) [reply zeebe/qa/util/src/main/java/io/camunda/zeebe/qa/util/cluster/TestCamundaApplication.java:14]: Will do.

File: dist/src/main/java/io/camunda/commons/clustering/AtomixClusterConfiguration.java (hunk 6 above)
Comment (romansmirnov) [reply dist/src/main/java/io/camunda/commons/clustering/AtomixClusterConfiguration.java:31]: * When the app gets started with the profile `broker`, the atomix cluster is started as one of the startup steps.
* When the app is started with the profile `gateway`, then the atomix cluster is started in the `GatewayModuleConfiguration` in the `dist` module.

Now, with that change, the atomix cluster gets started before the broker or the gateway actually starts it. Because, when the app is started with the profile `operate` then the broker and the gateway itself are not started, but the atomix cluster still must be started so that the app can join the cluster and forward v2 requests to the broker where applicable.

> How does this affect now the bootstrapping, is this an issue?
I went through the different paths; to me, they do not affect the bootstrapping. It starts the Atomix cluster, and if the broker once again tries to start it (when applying the startup steps), it's a noop because `start()` ensures that if it is already started, it won't be started again. Also, I couldn't spot any place where it must be started only after something different happened. Yes, the broker startup steps provide a specific order, but it doesn't there is nothing that relies on it. The same applies to the gateway startup procedure.

I checked if I could refactor that part so that it starts only in one place, but it seemed to require a bigger refactoring. So I went this path, and as I said, debugged the code paths and tested them locally but couldn't spot any issues.

File: dist/src/main/java/io/camunda/application/configuration/BrokerBasedConfiguration.java (hunk 3 above)
Comment (romansmirnov) [reply dist/src/main/java/io/camunda/application/configuration/BrokerBasedConfiguration.java:75]: Yes, that's what I understood from our discussion we are okay with that, and the user must disable the REST API explicitly. But maybe I misunderstood something 🤔 

In the first iteration, I kept it with `zeebe.broker.gateway.enable`; if set to `true`, it will start the REST API. That way, starting the broker would start the `gateway-grpc` and `gateway-rest`. If set to `false`, both would not be started. When running the app with another profile, it would disable the `gateway-rest` when set to `false`.

However, I could try to combine these properties, basically, if `zeebe.broker.gateway.enable` and `camunda.rest.enabled` set to `true`, then enable the REST API when the broker is started.

File: dist/src/main/java/io/camunda/commons/job/JobHandlerConfiguration.java (hunk 7 above)
Comment (romansmirnov) [reply dist/src/main/java/io/camunda/commons/job/JobHandlerConfiguration.java:32]: No, we are not breaking at this point. If the REST API is enabled, it just registers and initiates the job handler bean. Of course, if, in the case of the broker, the embedded gateway is disabled but the REST API enabled, it would create that bean, which, of course, is not nice, but it doesn't break anything because here, it does not expose any endpoints or anything. It just provides a bean that can be consumed by the REST API.

File: dist/src/main/java/io/camunda/application/configuration/BrokerBasedConfiguration.java (hunk 3 above)
Comment (romansmirnov) [reply dist/src/main/java/io/camunda/application/configuration/BrokerBasedConfiguration.java]: I adjusted the condition with 58cb2d9996b91f2a80f6e08800a0e92d93c0bdb2

When running the broker:
* `zeebe.broker.gateway.enable=false` then the embedded gateway is disabled, including the REST API
* `zeebe.broker.gateway.enable=true` and `camunda.rest.enabled=true` (default=`true`), then the embedded gateway is enabled, including the REST API
* `zeebe.broker.gateway.enable=true` and `camunda.rest.enabled=false, then the embedded gateway is enabled but the the REST API is disabled

When running without the broker
* `camunda.rest.enabled=true`, the REST API is started
* `camunda.rest.enabled=false`, the REST API is disabled
* When `zeebe.broker.gateway.enable` is set to `false`, the REST API is also disabled. However, if not present, the default value `true` is set, which enables the REST API.

An alternative would be implementing an initializer to handle the different cases when bootstrapping the app. But I felt this might be sufficient, and the last case seems to be an edge case.

File: dist/src/main/java/io/camunda/commons/CommonsModuleConfiguration.java (hunk 5 above)
Comment (romansmirnov) [reply dist/src/main/java/io/camunda/commons/CommonsModuleConfiguration.java:19]: I removed the `ApplicationModuleConfiguration` class, and moved the `configuration` into the `commons` package.

**Comment by romansmirnov** [comment]
@Zelldon, do you mind taking a look at the additional changes I made based on your feedback?

File: dist/src/main/java/io/camunda/commons/clustering/AtomixClusterConfiguration.java (hunk 6 above)
Comment (Zelldon) [reply dist/src/main/java/io/camunda/commons/clustering/AtomixClusterConfiguration.java:31]: Thanks for checking. 👍🏼 I remember some years ago we had issues with twice starting services (which would break, etc.) also depending on the order. But If you see no actual issue, I'm fine. but yes would be good to refactor this at somepoint.

File: dist/src/main/java/io/camunda/zeebe/restore/RestoreApp.java (hunk 9 above)
Comment (Zelldon) [reply dist/src/main/java/io/camunda/zeebe/restore/RestoreApp.java:29]: Thanks 👍🏼

File: dist/src/main/java/io/camunda/application/configuration/GatewayBasedConfiguration.java (hunk 4 above)
Comment (Zelldon) [reply dist/src/main/java/io/camunda/application/configuration/GatewayBasedConfiguration.java]: Thanks

**Review by ChrisKujawa — APPROVED** [review]
Great work thanks @romansmirnov 👍🏼

File: dist/src/main/java/io/camunda/commons/rest/RestApiConfiguration.java [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+package io.camunda.commons.rest;
+
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.context.annotation.ComponentScan;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration(proxyBeanMethods = false)
+@ComponentScan(basePackages = "io.camunda.zeebe.gateway.rest")
+@ConditionalOnProperty(
+    name = {"zeebe.broker.gateway.enable", "camunda.rest.enabled"},
```
[the remark this answers is not in the document]
Comment (romansmirnov) [reply dist/src/main/java/io/camunda/commons/rest/RestApiConfiguration.java:17]: Yep, exactly. It acts as an `and`.
